libraries/Microsoft.Bot.Builder.Dialogs/Prompts/CloudOAuthPrompt.cs

```diff
@@ -0,0 +1,137 @@
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        /// <remarks>If the task is successful, the result indicates whether the prompt is still
+        /// active after the turn has been processed by the prompt.</remarks>
+        public override Task<DialogTurnResult> ContinueDialogAsync(DialogContext dc, CancellationToken cancellationToken = default)
+        {
+            var userTokenClient = dc.Context.TurnState.Get<UserTokenClient>() ?? throw new InvalidOperationException("The UserTokenClient is not supported by the current adapter.");
+
+            // Check for timeout
+            if (OAuthHelper.HasTimeoutExpired(dc))
+            {
+                return dc.EndDialogAsync(null, cancellationToken);
+            }
+
+            if (dc.Context.Activity.Type == ActivityTypes.Message)
+            {
+                return OnContinueWithMessageActivityAsync(dc, userTokenClient, _settings.ConnectionName, _validator, _settings.EndOnInvalidMessage, cancellationToken);
+            }
+            else
+            {
+                return OAuthHelper.OnContinueWithNonMessageActivityAsync(dc, userTokenClient, _settings.ConnectionName, cancellationToken);
+            }
+        }
+
+        private static async Task<DialogTurnResult> OnContinueWithMessageActivityAsync(DialogContext dc, UserTokenClient userTokenClient, string connectionName, PromptValidator<TokenResponse> validator, bool endOnInvalidMessage, CancellationToken cancellationToken)
+        {
+            var tokenResponse = await OAuthHelper.CreateTokenResponseFromMessageAsync(userTokenClient, connectionName, dc.Context.Activity, cancellationToken).ConfigureAwait(false);
+
+            // Call any custom validation we might have.
+            if (await CheckValidatorAsync(dc, validator, tokenResponse, cancellationToken).ConfigureAwait(false))
+            {
+                return await dc.EndDialogAsync(tokenResponse, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                // If EndOnInvalidMessage is set, complete the prompt with no result.
+                if (endOnInvalidMessage)
+                {
+                    return await dc.EndDialogAsync(null, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // If this was a message Activity we can use the retry prompt
+                    var promptOptions = dc.ActiveDialog.State[OAuthHelper.PersistedOptions].CastTo<PromptOptions>();
+                    if (!dc.Context.Responded && promptOptions?.RetryPrompt != null)
+                    {
+                        await dc.Context.SendActivityAsync(promptOptions.RetryPrompt, cancellationToken).ConfigureAwait(false);
```

Comment:
Will this be okay, if the incoming activity type is not a 'message'?

Reply:
At this point the activity must be a message activity.

Reply:
oh, the method name :)